lib/ProductOpener/Nutrition.pm

```diff
@@ -308,20 +373,1471 @@ sub convert_nutrient_to_100g ($nutrient_ref, $original_per, $original_per_quanti
+		convert_nutrition_input_sets_hash_to_array($input_sets_hash_ref, $product_ref));
+
+	return;
+}
+
+=head2 assign_nutrition_values_from_imported_csv_product ( $imported_csv_product_ref, $product_ref, $nutriment_table )
+
+This function is used by Import.pm to import new nutrition data from an imported product (though a CSV file) to an existing product.
+
+It reads the new nutrition data parameters from the imported product, and assigns them to the new product nutrition structure.
+
+Note: the serving_size fields need to be imported first, as we need it to set the per_quantity and per_unit fields of the "serving" input sets.
+
+Note: a source is not specified as argument to this function, as it should be set in the field names.
+
+=head3 Parameters
+
+=head4 $imported_csv_product_ref
+
+Reference to the imported product hash where the nutrition data will be read.
+
+All the fields in the imported product are at the root level, e.g. nutrition.input_sets.as_sold.100g.nutrients.energy.value_string
+
+=head4 $product_ref
+
+Reference to the product hash where the nutrition data will be stored.
+
+=head4 $nutriment_table
+
+The nutriment table to use. It should be one of the keys of %nutriments_tables in Config.pm
+
+=head4 $source
+
+The source of the nutrition data. e.g. "packaging" or "manufacturer"
+
+=cut
+
+sub assign_nutrition_values_from_imported_csv_product ($imported_csv_product_ref, $product_ref) {
+
+	my @preparations = get_preparations_for_product_type($product_ref->{product_type});
+	my @pers = get_pers_for_product_type($product_ref->{product_type});
+
+	# We use a temporary input sets hash to ease setting values
+	my $input_sets_hash_ref = get_nutrition_input_sets_in_a_hash($product_ref);
+
+	# We identify all the sources included in the imported product fields
+	# We look for fields that start with nutrition.input_sets.${source}.${preparation}.${per}.nutrients.
+	my %sources = ();
+	foreach my $field (keys %{$imported_csv_product_ref}) {
+		if ($field =~ /^nutrition\.input_sets\.([a-zA-Z0-9_-]+)\.([a-z_]+)\.(100g|100ml|1kg|1l|serving)\.nutrients\./) {
+			my $source = $1;
+			$sources{$source} = 1;
+		}
+	}
+	my @sources = sort keys %sources;
+
+	# Assign all the nutrient values
+
+	foreach my $nutrient_tagid (sort(get_all_taxonomy_entries("nutrients"))) {
+
+		my $nid = $nutrient_tagid;
+		$nid =~ s/^zz://g;
+
+		next if $nid =~ /^nutrition-score/;
+
+		# nutrient values and units are passed for the different input sets (for each preparation type and for each per quantity)
+		# with parameters like:
+		#
+		# nutrition.input_sets.packaging.prepared.100ml.nutrients.saturated-fat.value_string
+		# nutrition.input_sets.packaging.prepared.100ml.nutrients.saturated-fat.unit
+		#
+		# Note: the parameters are long because they mimic the structure of the product nutrition hash
+
+		# Go through all the possible input sets
+		foreach my $preparation (@preparations) {
+			foreach my $per (@pers) {
+				foreach my $source (@sources) {
```

Comment:
it could be interesting to keep track of processed field, keep track of field that we consider while computing sources, and at the end, compare the two list to issue a warning about not processed field, it could help spot typos and so, that otherwise will be hard to debug.